internal/multistorage/cache/status_cache.go

```diff
@@ -21,169 +23,281 @@ type statusCache struct {
 	storagesInOrder []NamedFolder
 	ttl             time.Duration
 	checker         AliveChecker
+
+	// sharedMem keeps the intra-process cache that's shared between different threads and subsequent storages requests.
+	sharedMem   *storageStatuses
+	sharedMemMu *sync.Mutex
+
+	// sharedFile keeps the inter-process cache that's shared between different command executions.
+	sharedFile    string
+	sharedFileUse bool
+}
+
+type StatusCacheOpt func(c *statusCache)
+
+func WithSharedMemory(mem *storageStatuses, memMu *sync.Mutex) StatusCacheOpt {
+	return func(c *statusCache) {
+		c.sharedMem = mem
+		c.sharedMemMu = memMu
+	}
+}
+
+func WithSharedFile(file string) StatusCacheOpt {
+	return func(c *statusCache) {
+		c.sharedFile = file
+		c.sharedFileUse = true
+	}
+}
+
+func WithoutSharedFile() StatusCacheOpt {
+	return func(c *statusCache) {
+		c.sharedFileUse = false
+	}
 }
 
 func NewStatusCache(
 	primary storage.Folder,
 	failover map[string]storage.Folder,
 	ttl time.Duration,
 	checker AliveChecker,
+	opts ...StatusCacheOpt,
 ) (StatusCache, error) {
 	storagesInOrder, err := NameAndOrderFolders(primary, failover)
 	if err != nil {
 		return &statusCache{}, err
 	}
 
-	return &statusCache{
+	homeFileUse := true
+	homeFile, err := HomeStatusFile()
+	if err != nil {
+		tracelog.WarningLogger.Printf("Can't use storage status cache file from $HOME: %v", err)
+		homeFileUse = false
+	}
+
+	c := &statusCache{
 		storagesInOrder: storagesInOrder,
 		ttl:             ttl,
 		checker:         checker,
-	}, nil
+
+		sharedMem:     &globalMemCache,
+		sharedMemMu:   globalMemCacheMu,
+		sharedFile:    homeFile,
+		sharedFileUse: homeFileUse,
+	}
+
+	for _, o := range opts {
+		o(c)
+	}
+
+	return c, nil
 }
 
-func (c *statusCache) AllAliveStorages() ([]NamedFolder, error) {
-	memCacheMu.Lock()
-	defer memCacheMu.Unlock()
+func (c *statusCache) AllAliveStorages() (allAlive []NamedFolder, err error) {
+	c.sharedMemMu.Lock()
+	defer c.sharedMemMu.Unlock()
 
-	if memCache.isRelevant(c.ttl, c.storagesInOrder...) {
-		return memCache.getAllAlive(c.storagesInOrder), nil
+	if c.sharedMem.isRelevant(c.ttl, c.storagesInOrder...) {
+		allAlive = c.sharedMem.getAllAlive(c.storagesInOrder)
+		if len(allAlive) > 0 {
+			return allAlive, nil
+		}
 	}
 
-	oldFile, err := readFile()
-	if err != nil {
-		tracelog.WarningLogger.Printf("Failed to read cache file, it will be overwritten: %v", err)
+	var oldFile storageStatuses
+	if c.sharedFileUse {
+		oldFile, err = readFile(c.sharedFile)
+		if err != nil {
+			tracelog.WarningLogger.Printf("Failed to read cache file, it will be overwritten: %v", err)
+		}
 	}
-	_, outdated := oldFile.splitByRelevance(c.ttl, c.storagesInOrder)
+	relevant, outdated := oldFile.splitByRelevance(c.ttl, c.storagesInOrder)
```

Comment:
Will it result in nil dereference exception here if there is no shared file?

Reply:
No, and this situation (and all the similar ones) is covered by tests. All methods of `storageStatuses` struct can work if this is nil - because both memory or file could be nil.
In this specific case all storages will be considered "outdated".